src/unistd.rs

```diff
@@ -2136,3 +2146,368 @@ mod setres {
+    pub mem: Vec<String>
+}
+
+impl From<libc::group> for Group {
+    fn from(gr: libc::group) -> Group {
+        unsafe {
+            Group {
+                name: CStr::from_ptr(gr.gr_name).to_string_lossy().into_owned(),
+                gid: Gid::from_raw(gr.gr_gid),
+                mem: Group::members(gr.gr_mem)
+            }
+        }
+    }
+}
+
+impl Group {
+    unsafe fn members(mem: *mut *mut c_char) -> Vec<String> {
+        let mut ret = vec![];
+
+        for i in 0.. {
+            let u = mem.offset(i);
+            if (*u).is_null() {
+                break;
+            } else {
+                let s = CStr::from_ptr(*u).to_string_lossy().into_owned();
+                ret.push(s);
+            }
+        }
+
+        ret
+    }
+}
+
+#[derive(Debug)]
+/// Query a group. The buffer size variants should not be needed 99% of the time; the default
+/// buffer size of 1024 should be more than adequate. Only use the buffer size variants if you
+/// receive an ERANGE error without them, or you know that you will be likely to.
+pub enum UserQuery {
+    /// Get a user by UID. Internally, this function calls
+    /// [getpwuid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Uid(Uid),
+    /// Get a user by name. Internally, this function calls
+    /// [getpwnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    UidWithBufsize(Uid, usize),
+    NameWithBufsize(String, usize)
+}
+
+#[derive(Debug)]
+/// Query a user. The buffer size variants should not be needed 99% of the time; the default buffer
+/// size of 1024 should be more than adequate. Only use the buffer size variants if you receive an
+/// ERANGE error without them, or you know that you will be likely to.
+pub enum GroupQuery {
+    /// Get a group by GID. Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name. Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    GidWithBufsize(Gid, usize),
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
+}
+
+/// Default buffer size for system user and group querying functions
+const PWGRP_BUFSIZE: usize = 1024;
+
+impl Queryable<UserQuery> for User {
+    fn query(q: UserQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            UserQuery::UidWithBufsize(_, size) | 
+            UserQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+        let mut res = ptr::null_mut();
+
+        let i = unsafe { 
```

Comment:
Big `unsafe` blocks can easily hide bugs.  Better to use `unsafe` only around the code that needs it.

Reply:
this is changed in the latest version as well.